refc!: use new store, cleanup old files

gamemaker1 opened the pull request:
- Closes #5 by implementing the new store interface.
- ~Converts the package to pure ESM.~ Exports both ESM and CJS files.
- Upgrades all depedencies.

nfriedly commented on package.json:
```diff
@@ -1,78 +1,92 @@
+				"types": "./dist/index.d.mts",
+				"default": "./dist/index.mjs"
+			},
+			"require": {
+				"types": "./dist/index.d.cts",
+				"default": "./dist/index.cjs"
+			}
```
What do you think about dropping this and making rate-limit-memcached v2 an ESM-only module?

Might be a good place to test it out before doing the same with express-rate-limit.

gamemaker1 replied:
I think it's a good idea, let's do it!

nfriedly commented:
Thank you for doing this! I've started to look through it, but I'm not going to be able to finish reviewing tonight. Hopefully I can find some time this weekend.

nfriedly commented on source/store.ts:
```diff
@@ -0,0 +1,190 @@
+// /source/store.ts
```
I kind of preferred the old `memcached-store.ts` filename.

(Relatedly, I'd prefer if we renamed `lib.ts` to `express-rate-limit.ts` in that repo, but I didn't notice the name when I first reviewed it and never bothered to say anything afterwards.)

nfriedly commented on src/memcached-store.ts:
```diff
@@ -1,49 +0,0 @@
-
-export default class MemcachedStore {
-  public prefix: string
-  public expiration: number
```
Maybe we should keep this and add an `incr()` method to the new store for backwards-compatibility with older exress-rate-limit and current express-slow-down?

Also, then we could check in `init()` and verify that expiration matches windowMs/1000 and throw an error if they disagree.

nfriedly commented on source/store.ts:
```diff
@@ -0,0 +1,190 @@
+	async increment(key: string): Promise<IncrementResponse> {
+		const response = (await this.updateKey(
+			this.prefixKey(key),
+			+1,
```
`+1` looks funny here. Would `1` also work?

```suggestion
			1,
```

nfriedly commented on source/store.ts:
```diff
@@ -0,0 +1,190 @@
+		if (record === undefined && delta < 0) {
+			// If the record does not exist, and we are supposed to decrement the key,
+			// we don't need to do anything, so we return.
+			return
+		}
+
+		if (record === undefined) {
+			// If a record does not exist, and must be incremented, then add a record
+			// for that key.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { hits: 1, time: Date.now() }
+
+				this.client.add(key, payload, this.expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		} else {
+			// If it does exist, simply change the hit count by `delta`.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { ...record, hits: record.hits + delta }
+				const expiration = Math.floor(
+					// = window - (time elapsed since start of window) [in seconds]
+					this.expiration - (Date.now() - payload.time) / 1000,
+				)
+
+				this.client.replace(key, payload, expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		}
```
I liked the old flow of calling incr/decr. Calling get then create/update leaves it you open to race conditions where a bunch of requests come in at once, and all get the initial value then all overwrite eachother's additions.

IMO, removing this function and just letting increment and decrement each have their own logic probably makes more sense here.

gamemaker1 replied:
We could do that, but then we won't be able to return an accurate `resetTime` to the middleware.

nfriedly replied:
Oh, that's a good point. I should familiarize myself with memcahed's commands.

nfriedly replied:
What about `incr` followed by either creating the key if it's not found, or `mg` (meta get), with the `t` to return the expiration time in seconds if the `incr` worked.

It's still two commands, but there's no risk of collision (after the initial creation) this way.

gamemaker1 replied:
Oooh that sounds good, let me do that. Thanks!

gamemaker1 replied:
Maybe we should switch to https://github.com/electrode-io/memcache/tree/master/packages/memcache-client? It has support for sending raw commands, and is more actively maintained.

nfriedly replied:
Oh, I like that it returns promises also, so we can await it. Yeah, that seems like a good idea.

nfriedly replied:
It looks like you can use `memcached.command()` in the old memcached library to send arbitrary commands. It's documented as private, but it is documented.

It might be worth trying to build a compatibility layer to let existing installs continue to use the old memcache library, even if we don't recommend it for new installations. 

Somewhat similar to our legacy store wrapper in express rate limit.

nfriedly replied:
oh, snap,  the [Meta Arithmetic](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L851) command, `ma` is exactly what we want.

`ma mykey N60 v t` will:
 * create mykey if it doesn't exist with a expiration time of 60 seconds (`N60`)
 * increment it (`MI`, which is the default so I left it out) by 1 (`D1`, also default)
 * return the new value (`v`) and the expiration time (`t`)

That's a single memcached request per hit, regardless of weather or not the key already existed.

Similarly, `ma mykey N60 MD v t` will decrement (`MD`) it instead of incrementing.

I'm going to see if I can find a memcached client that supports meta commands, and I might send some PRs if I can't find one.

nfriedly replied:
I just filed a ticket on memcached-client requesting support for meta commands. I'm kind of afraid that I just signed myself up for a ton of work, but I think it will make this a better library in the long run.

https://github.com/electrode-io/memcache/issues/22

gamemaker1 replied:
> oh, snap, the [Meta Arithmetic](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L851) command, ma is exactly what we want.

It is!! Great discovery!

> https://github.com/electrode-io/memcache/issues/22

Thanks for opening the issue, I'll do my best to help implement it.

gamemaker1 replied:
I don't see any reply on the issue.. should we try pinging the maintainers?

Otherwise we could try using the private `.command` method on the library we are using right now.

nfriedly replied:
Yeah, might as well try that

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,106 @@
+	legacyHeaders: false, // Disable the `X-RateLimit-*` headers.
+	store: new MemcachedStore({
+		// prefix: 'rl:', // The prefix attached to all keys stored in the cache.
+		// client: new Memcached(['foo.bar', 'localhost:11211']) // The memcached client to use.
```
If we add support for passing in a locations param, then the example here can look like 
```suggestion
		locations: ['localhost:11211'], // server location(s), passed directly to Memcached
```

nfriedly reviewed (commented):
Nice work on this! I have a few thoughts below, the biggest ones are 

1. I think we have to bring cjs back
2. incr/decr flow
3. accept locations & options to pass to the memcached client

Then there's a bunch of smaller nitpicky things

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,106 @@
+# <div align="center"> Express Rate Limit - Memcached Store </div>
+
+<div align="center">
```
(I forgot we do this in express-rate-limit.. I guess it's fine.)

nfriedly commented on readme.md:
```diff
@@ -0,0 +1,106 @@
+[![npm version](https://img.shields.io/npm/v/rate-limit-memcached.svg)](https://npmjs.org/package/rate-limit-memcached 'View this project on NPM')
+[![npm downloads](https://img.shields.io/npm/dm/rate-limit-memcached)](https://www.npmjs.com/package/rate-limit-memcached)
+
+A memcached store for the
```
```suggestion
A [memcached](https://www.memcached.org/) store for the
```

gamemaker1 commented:
Thanks for the review comments, I'll address them by tomorrow!

gamemaker1 commented:
Made the suggested changes:

- updated copyright notice in the license and readme files
- used the lts version names in the workflow file
- added the `location` and `config` options.
- restored cjs support

nfriedly commented on source/memcached-store.ts:
```diff
@@ -0,0 +1,194 @@
+		if (record === undefined && delta < 0) {
+			// If the record does not exist, and we are supposed to decrement the key,
+			// we don't need to do anything, so we return.
+			return
+		}
+
+		if (record === undefined) {
+			// If a record does not exist, and must be incremented, then add a record
+			// for that key.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { hits: 1, time: Date.now() }
+
+				this.client.add(key, payload, this.expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		} else {
+			// If it does exist, simply change the hit count by `delta`.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { ...record, hits: record.hits + delta }
+				const expiration = Math.floor(
+					// = window - (time elapsed since start of window) [in seconds]
+					this.expiration - (Date.now() - payload.time) / 1000,
+				)
+
+				this.client.replace(key, payload, expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		}
```
I'd love to use the new meta arithmetic command, but for right now, what about this strategy:

1. call incr on the key, if it succeeds:
    * do a `get` on the key, suffixed with '_expiry'
    * return both values
2. if the incr call failed because the key doesn't exist:
    * set the key to 1
    * set `${key}_expiry` to the expiration time
    * return both of those values

This is roughly what the previous version did, except we're adding the expiration time in a new key. It also has the benefit of being backwards compatible with existing users.

We can always add the meta arithmetic version later as an enhancement later on.

nfriedly replied:
Also, I think that should go into the `increment` method, `decrement`should just call decr on the key (ignoring both the expiration time and any failures), and we should drop the `updateKey` method.

gamemaker1 replied:
I was working on the meta arithmetic command, but the private `.command` function doesn't seem to return the output from the redis command.

Your idea sounds good, I'll try implementing it.

nfriedly reviewed (commented):
Sorry this got sidelined for a bit. I think I posted a good plan to move forward below.

gamemaker1 commented on config/jest.json:
```diff
@@ -0,0 +1,13 @@
+	"moduleNameMapper": {
+		"^(\\.{1,2}/.*)\\.js$": "$1"
+	}
+}
```
👍🏾

gamemaker1 commented:
This should be ready to go!

nfriedly commented:
Awesome, I'll take another look sometime soon

nfriedly commented on source/memcached-store.ts:
```diff
@@ -0,0 +1,171 @@
+			totalHits = 1 // When you set it to 1, it returns `true` for some reason.
+
+			// Also store the expiration time in a separate key.
+			expiresAt = Date.now() + this.expiration
```
I think this is wrong; I think `Date.now()` returns a value in milliseconds, but `this.expiration` is in seconds because we divide by 1000 in `init()`

```suggestion
			expiresAt = Date.now() + (this.expiration * 1000)
```

nfriedly commented on source/memcached-store.ts:
```diff
@@ -0,0 +1,171 @@
+		// Return the total number of hits, as well as the reset timestamp.
+		return {
+			totalHits,
+			resetTime: expiresAt === undefined ? undefined : new Date(expiresAt),
```
I think undefined would only happen if it expired in between reading the hits and reading the expired time, right? So perhaps we could default it to "right now"?
```suggestion
			// if expiresAt is undefined, assume the key expired in between reading the hits and expiry keys from memcached
			resetTime: expiresAt ? new Date(expiresAt) : new Date(),
```

gamemaker1 replied:
Hmm makes sense

nfriedly commented on source/memcached-store.ts:
```diff
@@ -0,0 +1,171 @@
+		if (totalHits === false) {
+			// The increment command failed since the key does not exist. In which case, set the
+			// hit count for that key to 1, and make sure it expires after `window` seconds.
+			await setKey(prefixedKey, 1, this.expiration)
```
This technically leaves a window open where, if a bunch of requests came in at the same time, they'd each set the hits to 1, and only the last one would actually be counted. 

To block that, we could call `add` instead of `set` and then call increment if the `add` fails because the key already exists.

Requiring `add` on the `Store` API is a breaking change.

Not sure if it's worth the trouble, but we might just want to document it if we don't make the change.

gamemaker1 replied:
I think this qualifies as a bug, so I went ahead and changed the store to use `add` instead.

> Requiring add on the Store API is a breaking change.

👍🏾

nfriedly reviewed (changes requested):
Great work!

I left a ton of comments, but the `expiresAt` one is the only thing that looks like a real bug. The others are pretty minor.

nfriedly commented on source/memcached-store.ts:
```diff
@@ -0,0 +1,171 @@
+	 *
+	 * @returns {string} - The key + '__expiry'.
+	 */
+	expirySuffix(key: string): string {
+		return `${key}__expiry`
+	}
```
Oh, I probably should have said to add expiry to the prefix rather than the suffix, e.g. 
```suggestion
	expiryKey(key: string): string {
		return `${this.prefix}expiry:${key}`
	}
```

And then pass in the original key rather than the prefixed one.

It's not a big deal, but this feels a little cleaner.

gamemaker1 replied:
Done in commit 3899485.

gamemaker1 commented on package.json:
```diff
@@ -1,78 +1,92 @@
-    "prepublishOnly": "npm run clean && npm run build && npm test"
-  }
+	"name": "rate-limit-memcached",
+	"version": "1.0.0",
```
> I see that package.json says 1.0.0 right now, but it looks like that never actually got shipped. Lets revert this to 0.6.0 here, and then when we're ready to publish, we'll do npm version major to bump it back to 1.x with a git tag.

Ooh I didn't notice that, will do.

gamemaker1 commented on readme.md:
```diff
@@ -0,0 +1,130 @@
+> Here, `key` is a string, `value` is a number, and `lifetime` is the time in
+> seconds until it expires.
+
+Defaults to [`new Memcached()`](https://github.com/3rd-Eden/memcached).
```
I went with:

```md
Defaults to an instance of [`memcached`](https://github.com/3rd-Eden/memcached),
created with the `locations` and `config` options (see below for details) passed
to it.
```

Does that seem ok?

gamemaker1 commented:
I think I addressed all the review comments, let me know if I should change anything else.

nfriedly reviewed (approved):
Nice work!
